[WebGPU] forceFallbackAdapter CTS test fails (251629)

mwyrzykowski opened the pull request:
#### 688149b429be877859932aecb386b215a2c60717
<pre>
[WebGPU] forceFallbackAdapter CTS test fails (251629)
<a href="https://bugs.webkit.org/show_bug.cgi?id=251629">https://bugs.webkit.org/show_bug.cgi?id=251629</a>
&lt;radar://104976449&gt;

Reviewed by Myles C. Maxfield.

The spec says to return nullptr from requestAdapter if the
forceFallback adapter is not supported.

Now we get 0 failures (3 skips, 6 passes), on the requestAdapter
CTS tests.

* Source/WebCore/Modules/WebGPU/GPU.cpp:
(WebCore::GPU::requestAdapter):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUImpl.cpp:
(PAL::WebGPU::GPUImpl::requestAdapter):
* Source/WebGPU/WebGPU/Adapter.mm:
(WebGPU::Adapter::requestDevice):
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::requestAdapter):
(wgpuInstanceRequestAdapter):
(wgpuInstanceRequestAdapterWithBlock):

Canonical link: <a href="https://commits.webkit.org/259844@main">https://commits.webkit.org/259844@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/263717910740905b11f128918f144836bf0239c2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106112 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15165 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115296 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175373 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16597 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6337 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98321 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114993 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111868 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95598 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94494 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27241 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8416 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28593 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8910 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/5141 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14529 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48137 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6810 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10455 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/Adapter.mm:
```diff
@@ -85,33 +85,25 @@
-        instance().scheduleWork([strongThis = Ref { *this }, callback = WTFMove(callback)]() mutable {
-            callback(WGPURequestDeviceStatus_Error, Device::createInvalid(strongThis), "Unknown descriptor type"_s);
-        });
+        callback(WGPURequestDeviceStatus_Error, Device::createInvalid(*this), "Unknown descriptor type"_s);
```
I think this is backwards, and we should make requestAdapter() and requestDevice properly async (now that we can).

This change is probably okay if it's temporary, but let's at least file a bug about making those functions actually async.

(And ditto for below.)

mwyrzykowski replied:
I've filed https://bugs.webkit.org/show_bug.cgi?id=251713 and will update the FIXME that exists below.

But in the case of failure, there is no reason the callback should be called async, we can return immediately and notify the client of the failure. This is inline with the web and native specifications.

I can understand the reasoning for making the callback asynchronous if the device creation takes a while, but then it should probably occurring off the main thread in the case of native webgpu

litherum replied:
> But in the case of failure, there is no reason the callback should be called async, we can return immediately and notify the client of the failure.

The [spec says](https://gpuweb.github.io/gpuweb/#dom-gpuadapter-requestdevice):
> 1. Let contentTimeline be the current [Content timeline](https://gpuweb.github.io/gpuweb/#content-timeline).
> 2. Let promise be [a new promise](https://webidl.spec.whatwg.org/#a-new-promise).
> 3. Let adapter be this.[[[adapter]]](https://gpuweb.github.io/gpuweb/#dom-gpuadapter-adapter-slot).
> 4. Issue the initialization steps to the [Device timeline](https://gpuweb.github.io/gpuweb/#device-timeline) of this.
> 5. Return promise.

Because validation happens on the device timeline, I think failure is still supposed to be reported asynchronously.

> I can understand the reasoning for making the callback asynchronous if the device creation takes a while, but then it should probably occurring off the main thread in the case of native webgpu

`MTLDevice`s are singletons, so creating one is not slow. `requestDevice()` does more than that, though: it sorts and filters devices, inspects them for family capabilities, and, creates a queue. It sounds worthy of a new bug to determine whether or not we get benefit by doing device creation on a secondary thread.

mwyrzykowski replied:
Maybe I misunderstand but aren't the concepts of device and content timelines specific to the web api and not the native api?

A native webgpu application could have everything on the main thread if it desired, in which case the two timelines are the same.

litherum replied:
I think this is the same discussion as https://github.com/WebKit/WebKit/pull/9584#discussion_r1096406506 - in the WebGPU API, it's impossible for the promise's `.then` callback to be called in the middle of `requestAdapter()`. That's the semantics I'm trying to preserve.

mwyrzykowski committed:
[WebGPU] forceFallbackAdapter CTS test fails (251629)
https://bugs.webkit.org/show_bug.cgi?id=251629
<radar://104976449>

Reviewed by Myles C. Maxfield.

The spec says to return nullptr from requestAdapter if the
forceFallback adapter is not supported.

Now we get 0 failures (3 skips, 6 passes), on the requestAdapter
CTS tests.

* Source/WebCore/Modules/WebGPU/GPU.cpp:
(WebCore::GPU::requestAdapter):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUImpl.cpp:
(PAL::WebGPU::GPUImpl::requestAdapter):
* Source/WebGPU/WebGPU/Adapter.mm:
(WebGPU::Adapter::requestDevice):
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::requestAdapter):
(wgpuInstanceRequestAdapter):
(wgpuInstanceRequestAdapterWithBlock):

Canonical link: https://commits.webkit.org/259844@main
